pkg/jobrunaggregator/jobruntestcaseanalyzer/analyzer.go

```diff
@@ -0,0 +1,404 @@
+// 1. either gets a list of jobs from big query that meet the passed criteria: platform, network, or uses the passed jobs
+// 2. finds job runs for matching jobs for the specified payload tag
+// 3. runs all test case checkers and constructs a synthetic junit
+type JobRunTestCaseAnalyzerOptions struct {
+	payloadTag          string
+	workingDir          string
+	jobRunStartEstimate time.Time
+	clock               clock.Clock
+	timeout             time.Duration
+	platform            string
+	infrastructure      string
+	network             string
+	ciDataClient        jobrunaggregatorlib.CIDataClient
+	ciGCSClient         jobrunaggregatorlib.CIGCSClient
+	gcsClient           *storage.Client
+	testCaseCheckers    []TestCaseChecker
+	payloadInvocationID string
+	jobGCSPrefixes      *[]jobGCSPrefix
+}
+
+func getJobInfrastructure(name string) string {
+	if strings.Contains(name, "upi") {
+		return "upi"
+	}
+	return "ipi"
+}
+
+func (o *JobRunTestCaseAnalyzerOptions) filterJobsForPayload(allJobs []jobrunaggregatorapi.JobRow) []jobrunaggregatorapi.JobRow {
+	jobs := []jobrunaggregatorapi.JobRow{}
+	for i := range allJobs {
+		job := allJobs[i]
+		if (len(o.platform) != 0 && job.Platform != o.platform) ||
+			(len(o.network) != 0 && job.Network != o.network) ||
+			(len(o.infrastructure) != 0 && o.infrastructure != getJobInfrastructure(job.JobName)) {
+			continue
+		}
+		jobs = append(jobs, job)
+	}
+	return jobs
+}
+
+func (o *JobRunTestCaseAnalyzerOptions) filterJobsByNames(jobNames []string, allJobs []jobrunaggregatorapi.JobRow) ([]jobrunaggregatorapi.JobRow, error) {
+	jobMap := map[string]jobrunaggregatorapi.JobRow{}
+	for i := range allJobs {
+		job := allJobs[i]
+		jobMap[job.JobName] = job
+	}
+	jobs := []jobrunaggregatorapi.JobRow{}
+	for _, jobName := range jobNames {
+		if job, ok := jobMap[jobName]; !ok {
+			return jobs, fmt.Errorf("invalid job %s", jobName)
+		} else {
+			jobs = append(jobs, job)
+		}
+	}
+	return jobs, nil
+}
+
+// getRelatedJobs find all related jobs for the test case analyzer
+// For PR payload, this contains jobs correspond to the list of jobGCSPreix passed
+// For release-controller generated payload, this contains all jobs meeting selection criteria
+// from command args.
+func (o *JobRunTestCaseAnalyzerOptions) getRelatedJobs(ctx context.Context) ([]jobrunaggregatorapi.JobRow, error) {
+	jobs, err := o.ciDataClient.ListAllJobs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all jobs: %w", err)
+	}
+	// if PR payload, only find the exact jobs
+	if len(o.payloadInvocationID) > 0 {
+		jobNames := []string{}
+		for i := range *o.jobGCSPrefixes {
+			jobGCSPrefix := (*o.jobGCSPrefixes)[i]
+			jobNames = append(jobNames, jobGCSPrefix.jobName)
+		}
+		jobs, err = o.filterJobsByNames(jobNames, jobs)
+		if err != nil {
+			return nil, fmt.Errorf("failed to filter jobs by names: %w", err)
+		}
+	} else {
+		// Non PR payload, select by criteria
+		jobs = o.filterJobsForPayload(jobs)
```

Comment:
We do not filter by release. Different payload runs have different payload invocation ID. For jobs not specifically for that release, it will not get any job runs by using that invocation ID.